### ISSUE TITLE
Deployment created by Function CRD with service account annotation does not automount the token

### ISSUE THREAD

**Issue by kevin-lindsay-1** [openfaas/faas-netes#892]
In kubernetes, a service account with `automountServiceAccountToken: false` does not automatically mount the token into pods created by the deployment.

In order to accommodate this, you need to create a service account which has `automountServiceAccountToken: true` if you'd like a pod to be able to use it.

Instead, I would think it's a more advisable security practice to instead create a service account with `automountServiceAccountToken: false` and then in the deployment created by OpenFaaS, set `automountServiceAccountToken: true` for the pod.

## Expected Behaviour

Normally when I create service accounts I set automount to false, and then set automount to true on the pods that will need it.

## Current Behaviour

Right now pods do not specify that they would like a token automounted, so if you don't make the service account automount itself, the pod will not attempt to mount it.

## Are you a GitHub Sponsor (Yes/No?)

Check at: https://github.com/sponsors/openfaas
- [x] Yes
- [ ] No

## List All Possible Solutions and Workarounds

* You could just ignore it, as a service account should generally be on a per-function basis. But, I'd say automounting off is a perfectly reasonable setting.
* You could make a function set `automountServiceAccountToken` to `true` when there is a value for `com.openfaas.serviceaccount`.

Whether or not functions should even have a default service account, and that it should be automounted by default is a different question.

## Which Solution Do You Recommend?

The latter. Automount being off by default for a service account, and then opted-in specifically by pods seems easier to reason about, as well as my gut tells me there's possible security smells from having automount on all service accounts in a namespace by default.

## Steps to Reproduce (for bugs)

1. Create a service account with automount set to false
2. There is no way to configure the pod to use the service account

## Context

Trying to make a function that uses the k8s api.

## Your Environment

* FaaS-CLI version ( Full output from: `faas-cli version` ):

* Docker version `docker version` (e.g. Docker 17.0.05 ):

* What version and distriubtion of Kubernetes are you using? `kubectl version`

* Operating System and version (e.g. Linux, Windows, MacOS):

* Link to your project or a code example to reproduce issue:

* What network driver are you using and what CIDR? i.e. Weave net / Flannel

**Comment by kevin-lindsay-1** [comment]
@LucasRoesler I merely propose that the deployment created should at a pod level set automount on if there's a service account associated with it.

Right now, you have to create the service account with automount enabled for all pods that reference it.

There's 4 levels:
1. default sa auto mount
2. sa auto mount
3. pod auto mount
4. pod manual mount

Pod manual mount requires application-level code to do something that's a 1-line code change in k8s to enable, which is inconvenient and possibly error-prone (and kind of unnecessary in the first place), so I suggest targeting level 3 and leave levels 1 and 2 turned off, if possible. This way service accounts don't control automounting, and instead the pod requests it to be automounted.

Right now, you have to enable automount at level 2 if you don't want to deal with the hassle of doing it at level 4, even though you're really setting the token at level 3. The deployment generated just doesn't do that, probably just because it was a forgotten implementation detail.

**Comment by LucasRoesler** [comment]
@kevin-lindsay-1 my understanding of your level (3) is that the OpenFaaS deployment should _always_ enable automount on the Pod level.  I propose that this is actually undesirable. This would result in the pod _always_ mounting the token regardless of the ServiceAccount configuration. I don't think we can justify OpenFaaS overriding the configuration that the cluster admin set when they created the ServiceAccount, especailly because we have no sense of the level of cluster authorization the deployer has. 

My other proposal is that with options (1) and (2) you can achieve _all_ of the desired capabilities for a function and respect cluster level RBAC permissions.  You want to operate at layer (3), but to me that looks equivalent to layer (2) in which you copy or create a new ServiceACcount and toggle the automount value to the desired state (on or off). 

Especially because the use case we are discussing already has a non-default ServiceAccount, deploying the function already assumes that someone has created or can create the custom ServiceAccount. Hence they can also create this ServiceAccount with the correct automount value. 

I still don't understand the context to what capability or feature is blocked by this with the existing workarounds (copy and toggle at layer 2)? Especially when we consider the use case of unprivileged users being allowed to deploy via OpenFaaS, why should it override the ServiceAccount config, especially on an object that is not owned and managed by OpenFaaS?  With a good story for why it is needed and when it should be used, i would be pretty happy to support a new annotation.

I am very curious to hear @alexellis's ideas as well.

**Comment by kevin-lindsay-1** [comment]
> my understanding of your level (3) is that the OpenFaaS deployment should always enable automount on the Pod level

no, just the ability to specify. just asking for an API, not an opinionated solution.

> I still don't understand the context to what capability or feature is blocked by this with the existing workarounds (copy and toggle at layer 2)?

Nothing is blocked, this is just a best practice and exposition of pretty basic k8s functionality that admins are used to.

> why should it override the ServiceAccount config

Because otherwise anything that references the service account at all can mount the token.

---

As far as I'm concerned, level 2 is already implemented. If I automount the token and the pod references it, the pod automounts the token 100% of the time.

I was referring to the pod specifically opting into automounting it, rather than automount by reference alone. That, and if you want a pod to be able to automount a service account at level 2 there's no way to have another pod manual mount (for whatever reason). Since you can't opt-in right now you also can't opt-out.

Most people recommend turning automount off at the sa level and instead turning it on at the pod/deployment level. That's what I'm suggesting. Whether or not openfaas uses a boolean to say whether or not it would like a pod to automount isn't really an issue.

The issue was that if I made a sa I had to enable automount at the sa level, because you can't enable it at the pod level right now. I was merely proposing the ability to opt-in on a pod level, as is the best practice from what I can tell.

So I'm merely asking for the ability to flip a switch that tells the pod to automount or not. The config of the SA is completely out of scope for this IMO, as

**Comment by kevin-lindsay-1** [comment]
It's probably a good idea, yes.

In the truth table provided, the only way to mount said SA token with OF is to use case 2, where all pods automount a SA token. In reality, the best practice is likely case 5, where only the pods that need them automount.

You would probably be able to simply specify a SA for a function via an annotation and it automounts that token, rather than needing an annotation to specify the SA and another to enable it.
